File: DevelopArchitecture/DevelopArchitecture/threeclass/photo/JPhotosPreviewsVC.m
//
//  JPhotosPreviewsVC.m
//  DevelopArchitecture
//
//  Created by Jeffrey on 2019/4/2.
//  Copyright © 2019 Jeffrey. All rights reserved.
//

#import "JPhotosPreviewsVC.h"
#import "JPhotosPreviewsCell.h"
#import "JPhotoPickerInterface.h"
@interface JPhotosPreviewsVC ()<UICollectionViewDelegate>
@property(nonatomic,assign)BOOL m_isOrigalImage;

@end

@implementation JPhotosPreviewsVC
maddProperyValue(m_selectDataArr, NSMutableArray);

- (void)viewDidLoad {
    [super viewDidLoad];
    [JPhotoPickerInterface funj_addConfigSubView:self];
     [self.m_defaultImageView removeFromSuperview];
    UICollectionViewFlowLayout*flowfayout =(UICollectionViewFlowLayout*)self.m_collectionView.collectionViewLayout;
    flowfayout.minimumLineSpacing=0;
    flowfayout.scrollDirection = UICollectionViewScrollDirectionHorizontal;
    flowfayout.minimumInteritemSpacing=0;
    self.m_collectionView.pagingEnabled = YES;
    [self.m_collectionView registerClass:[JPhotosPreviewsCell class] forCellWithReuseIdentifier:kCellIndentifier];
    [self funj_addSubBottomBgView];
    self.m_collectionView.alwaysBounceVertical = NO;
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.3 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        if(self.m_scrollIndex > 0 && self.m_scrollIndex <self.m_dataArray.count){
            [self.m_collectionView scrollToItemAtIndexPath:[NSIndexPath indexPathForRow:self.m_scrollIndex inSection:0] atScrollPosition:UICollectionViewScrollPositionCenteredHorizontally animated:NO];
        }
    });
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [self scrollViewDidEndDecelerating:self.m_collectionView];
    });
}
-(void)funj_addSubBottomBgView{
    self.navigationItem.rightBarButtonItem =[UIBarButtonItem funj_getNavPublicButton:self  t:nil  a:@"funj_selectToAdd:" img:@"photo_def_photoPickerVc" set:^(UIButton *button) {
        [button setImage:[UIImage imageNamed:@"photo_sel_photoPickerVc"] forState:UIControlStateSelected];
        [button funj_updateButtonSelectStyle:NO  ischange:NO];
     }];
    
    UIView *bottomBgView =[UIView funj_getView:CGRectMake(0, self.view.height-50, self.view.width, 50) bg:kColor_White_Dark];
    [self.view addSubview:bottomBgView];bottomBgView.tag = 3023;
    
    UIImageView *line =[UIImageView funj_getLineImageView:CGRectMake(0, 0, kWidth, 1)];
    [bottomBgView addSubview:line];
    
    if(![JPhotosConfig share].m_currentIsVideo){
        UIButton *origareBt =[UIButton funj_getButtons:CGRectMake(0, 0, 100, 50) t:LocalStr(@"Original") fc:JTextFCMake(kFont_Size13, kColor_Text_Black_Dark) img:@[@"photo_original_def",@"photo_original_sel"] d:self  a:@"funj_selectItemTo:" tag:3024 set:^(UIButton *button) {
            [button funj_updateContentImageLayout:kLEFT_IMAGECONTENT a:JAlignMake(10, 10, 0)];
            [button funj_updateButtonSelectStyle:NO  ischange:NO];
        }];
        [bottomBgView addSubview:origareBt];
    }
    
    UIButton *sumBt =[UIButton funj_getButtons:CGRectMake(self.view.width-100, 0, 100, 50) t:LocalStr(@"Confirm") fc:JTextFCMake(kFont_Size13, kColor_Text_Black_Dark) img: nil d:self  a:@"funj_selectFinishTo:" tag:3025 set:^(UIButton *button) {
        [button funj_updateContentImageLayout:kRIGHT_CONTENTIMAGE a:JAlignMake(10, 0, 20)];
    }];
    [bottomBgView addSubview:sumBt];
}

-(NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section{
  return [self.m_dataArray count];
}
-(UICollectionViewCell*)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath{
     JPhotosPreviewsCell *cell=[collectionView dequeueReusableCellWithReuseIdentifier:kCellIndentifier forIndexPath:indexPath];
    self.m_defaultImageView.hidden = YES;
    if(self.m_scrollIndex == -1 || indexPath.row == self.m_scrollIndex){
        self.m_scrollIndex = -1;
        [cell funj_setBaseCollectionData:self.m_dataArray[indexPath.row] ];
    }
    return cell;
}
-(void)collectionView:(UICollectionView *)collectionView willDisplayCell:(UICollectionViewCell *)cell forItemAtIndexPath:(NSIndexPath *)indexPath{
    for(JPhotosPreviewsCell *cell in self.m_collectionView.visibleCells){
        [cell funj_stopAllPlayer];
    }
}
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath  {
    return CGSizeMake(collectionView.width,self.m_collectionView.height);
}

- (BOOL)popoverPresentationControllerShouldDismissPopover:(UIPopoverPresentationController *)popoverPresentationController{
    return NO;
}
-(void)viewDidLayoutSubviews{
    [super viewDidLayoutSubviews];
    [self funj_reloadBaseViewParameter:CGRectZero f:CGRectMake(0, 0, self.view.width, self.view.height-50) hidden:YES];
    UIView *bottomBgView =[self.view viewWithTag:3023];
    bottomBgView.top = self.view.height-50;
    bottomBgView.width = self.view.width;
    UIButton *sumBt =[bottomBgView viewWithTag:3025];
    sumBt.left = bottomBgView.width - 100;
}

-(void)funj_selectItemTo:(UIButton*)sender{
    sender.selected = !sender.selected;
    self.m_isOrigalImage = sender.selected;
}

-(void)funj_selectFinishTo:(UIButton*)sender{
    if(self.m_selectDataArr.count<=0){
        if(self.m_selectDataArr.count >= [JPhotosConfig share].m_maxCountPhotos){
            return;
        }
        UIButton *sender =[self.navigationItem.rightBarButtonItem customView];
        [self funj_selectToAdd:sender];
        if(self.m_selectDataArr.count<=0)return;
    }

    __block NSMutableDictionary *saveImageDic =[[NSMutableDictionary alloc]init];
    [self funj_showProgressView:nil];
    LRWeakSelf(self);
    for(NSInteger i =0;i<self.m_selectDataArr.count;i++){
        JPhotoPickerModel*model =self.m_selectDataArr[i];
        if([JPhotosConfig share].m_currentIsVideo){
            [JPhotoPickerInterface funj_getVideoWithAsset:model.asset completion:^(AVPlayerItem * _Nullable item, NSDictionary * _Nullable dic) {
                LRStrongSelf(self);
                dispatch_main_async_safe(^{
                    [self funj_getDetailInfo:saveImageDic img:nil d:dic  isDegrad:YES item:item index:i];
                });
            }];
        }else{
            [JPhotoPickerInterface funj_getPhotoWithAsset:model.asset type:PHImageRequestOptionsDeliveryModeHighQualityFormat photoWidth:kWidth completion:^(UIImage * _Nonnull image, NSDictionary * _Nonnull dic, BOOL isDegraded) {
                LRStrongSelf(self);
                if (isDegraded) {
                    return;
                }
                dispatch_main_async_safe(^{
                    [self funj_getDetailInfo:saveImageDic img:image d:dic  isDegrad:isDegraded item:nil index:i];
                });
            }];
        }
    }
}
-(void)funj_getDetailInfo:(NSMutableDictionary*)saveImageDic img:(UIImage*)image d:(NSDictionary*)dic isDegrad:(BOOL) isDegraded item:(AVPlayerItem*)item index:(NSInteger)index{
    if([JPhotosConfig share].m_currentIsVideo){
        if(item)[saveImageDic setObject:item forKey:@(index)];
    }else{
        if(image)[saveImageDic setObject:image forKey:@(index)];
    }
    if(saveImageDic.count == self.m_selectDataArr.count){
        [self funj_closeLoadingProgressView];
        NSMutableArray *saveArr =[[NSMutableArray alloc]init];
        for(int j=0;j<saveImageDic.count;j++){
            if([saveImageDic objectForKey:@(j)])
                [saveArr addObject:[saveImageDic objectForKey:@(j)]];
        }
        if([JPhotosConfig share].m_selectCallback){
            [JPhotosConfig share].m_selectCallback(saveArr, [JPhotosConfig share].m_currentIsVideo);
        }
        self.m_currentShowVCModel = kCURRENTISPRENTVIEW;
        self.modalTransitionStyle = UIModalTransitionStyleCrossDissolve;
        [self funj_clickBackButton:nil];
    }
}
-(void)funj_selectToAdd:(UIButton*)sender{
    sender.selected = !sender.selected;
    if(sender.selected && self.m_selectDataArr.count >= [JPhotosConfig share].m_maxCountPhotos){
        sender.selected = NO;
        sender.selected = NO;
        [JAppViewTools funj_showTextToast:self.view msg:[NSString stringWithFormat:LocalStr(@"You can only choose %zd photos"),[JPhotosConfig share].m_maxCountPhotos]];
        return ;
    }
    int index = (self.m_collectionView.contentOffset.x+10)/self.m_collectionView.width;
    if(index<self.m_dataArray.count){
        JPhotoPickerModel*model =self.m_dataArray[index];
        if(sender.selected){
            model.isSelected = YES;
            [self.m_selectDataArr addObject:model];
        }else{
            model.isSelected = NO;
            [self.m_selectDataArr removeObject:model];
        }
    }
}
-(void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView{
    int index = (scrollView.contentOffset.x+10)/scrollView.width;
    UIButton *rightBt = [self.navigationItem.rightBarButtonItem  customView];
    rightBt.selected = NO;
    if(index<self.m_dataArray.count){
        JPhotoPickerModel*model =self.m_dataArray[index];
        if([self.m_selectDataArr containsObject:model]){
            rightBt.selected = model.isSelected;
        }
    }
 
}
-(void)scrollViewDidEndScrollingAnimation:(UIScrollView *)scrollView{
    [self scrollViewDidEndDecelerating:scrollView];
}
-(void)funj_clickBackButton:(UIButton *)sender{
     if(self.m_changeCallback)self.m_changeCallback(self.m_selectDataArr, self.m_isOrigalImage);
    [super funj_clickBackButton:sender];
}
@end

File: DeveloperArchitecture/DeveloperArchitecture/threeclass/photo/JPhotosPreviewsVC.swift
//
//  JPhotosPreviewsVC.swift
//  DeveloperArchitecture
//
//  Created by Jeffrey on 2020/11/7.
//  Copyright © 2020 Jeffery. All rights reserved.
//

import Foundation
import UIKit
import AVFoundation

typealias kchangeCallback = (_ dataArr : [Any]? , _ isOrigalImage : Bool) -> ()

class JPhotosPreviewsVC : JBaseCollectionVC {
    var m_scrollIndex : Int = -1
    var m_selectDataArr : [Any]? = []
    var m_isOrigalImage : Bool = false
    var m_changeCallback : kchangeCallback?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        JPhotoPickerInterface.funj_addConfigSubView(self)
        self.m_defaultImageView.removeFromSuperview()
        let flowlayout = self.m_collectionView.collectionViewLayout as? UICollectionViewFlowLayout
        flowlayout?.minimumLineSpacing = 0
        flowlayout?.minimumInteritemSpacing = 0
        flowlayout?.scrollDirection = .horizontal
        self.m_collectionView.isPagingEnabled = true
        
        self.m_collectionView.register(JPhotosPreviewsCell.self, forCellWithReuseIdentifier: kCellIndentifier)
        self.funj_addSubBottomBgView()
        self.m_collectionView.alwaysBounceVertical = false
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) { [weak self] in
            if (self?.m_scrollIndex ?? 0) > 0 && (self?.m_scrollIndex ?? 0) < (self?.m_dataArray.count ?? 0) {
                self?.m_collectionView.scrollToItem(at: IndexPath(row: (self?.m_scrollIndex ?? 0), section: 0), at: .centeredHorizontally, animated: false)
            }
        }
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) { [weak self] in
            self?.scrollViewDidEndDecelerating(self!.m_collectionView)
        }
    }
    func funj_addSubBottomBgView() {
        self.navigationItem.rightBarButtonItem = UIBarButtonItem(i: nil , image: "photo_def_photoPickerVc", setButton: { (button) in
            button.funj_add(autoSelect: false)
            .setImage(UIImage(named: "photo_sel_photoPickerVc"), for: .selected)

        }, target: self, action: "funj_selectToAdd:")
        
        let bottomBgView = UIView(i: CGRect(x: 0, y: self.view.height - 50, width: self.view.width, height: 50), bg: kColor_White_Dark)
        self.view.addSubview(bottomBgView) ; bottomBgView.tag = 3023
        
        let line = UIImageView(i_line: CGRect(x: 0, y: 0, width: kWidth, height: 1))
        bottomBgView.addSubview(line)
        
        if JPhotosConfig.shared?.m_currentIsVideo ?? false {
            let origareBt = UIButton(i: CGRect(x: 0, y: 0, width: 100, height: 50), title: kLocalStr("Original"), textFC: JTextFC(f: kFont_Size13, c: kColor_Text_Black_Dark))
                .funj_add(bgImageOrColor: ["photo_original_def","photo_original_sel"], isImage: true)
                .funj_add(targe: self, action: "funj_selectItemTo:", tag: 3024)
                .funj_add(autoSelect: false)
            bottomBgView.addSubview(origareBt)
        }
        let sumBt = UIButton(i: CGRect(x: self.view.width - 100, y: 0, width: 100, height: 50), title: kLocalStr("Confirm"), textFC: JTextFC(f: kFont_Size13, c: kColor_Text_Black_Dark))
            .funj_add(targe: self, action: "funj_selectFinishTo:", tag: 3025)
            .funj_updateContentImage(layout: .kRIGHT_CONTENTIMAGE, a: JAlignValue(h: 10, s: 0, f: 20))
        bottomBgView.addSubview(sumBt)
    }
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        self.funj_reloadTableView(CGRectZero, table: CGRect(x: 0, y: 0, width: self.view.width, height: self.view.height - 50))
        let bottomBgView = self.view.viewWithTag(3023)
        bottomBgView?.top = self.view.height - 50
        bottomBgView?.width = self.view.width
        let sumBt = bottomBgView?.viewWithTag(3025)
        sumBt?.left = (bottomBgView?.width ?? 0) - 100
    }
}
extension JPhotosPreviewsVC {
    @objc func funj_selectItemTo (_ sender : UIButton) {
        sender.isSelected = !sender.isSelected
        self.m_isOrigalImage = sender.isSelected
    }
    @objc func funj_selectFinishTo (_ sender : UIButton) {
        if self.m_selectDataArr?.count ?? 0 <= 0 {
            if self.m_selectDataArr?.count ?? 0 >= (JPhotosConfig.shared?.m_maxCountPhotos ?? 0) {return }
            let sender = self.navigationItem.rightBarButtonItem!.customView as! UIButton
            self.funj_selectToAdd(sender)
            if self.m_selectDataArr?.count ?? 0 <= 0 { return }
        }
        self.funj_showProgressView()

        var saveImageDic : [AnyHashable : Any] = [:]

        for i in 0..<(self.m_selectDataArr?.count ?? 0) {
            let model = self.m_selectDataArr?[i] as? JPhotoPickerModel
            if JPhotosConfig.shared?.m_currentIsVideo ?? false {
                JPhotoPickerInterface.funj_getVideoWithAsset(asset: model?.m_asset) { [weak self ](item , dic) in
                    DispatchQueue.main.asyncAfter(deadline: .now() + 0.01) {
                        self?.funj_getDetailInfo(&saveImageDic, image: nil, dic: dic, isDegraded: true, item: item, index: i)
                    }
                }
            } else {
                JPhotoPickerInterface.funj_getPhotoWithAsset(phAsset: model?.m_asset, deliveryMode: .highQualityFormat, width: kWidth) { [weak self] (image, dic , isDegraded) in
                    if isDegraded { return }
                    DispatchQueue.main.asyncAfter(deadline: .now() + 0.01) {
                        self?.funj_getDetailInfo(&saveImageDic, image: image, dic: dic , isDegraded: isDegraded, item: nil, index: i)
                    }
                }
            }
            
        }
    }
    @objc func funj_selectToAdd (_ sender : UIButton) {
        sender.isSelected = !sender.isSelected
        if sender.isSelected && (self.m_selectDataArr?.count ?? 0) >= (JPhotosConfig.shared?.m_maxCountPhotos ?? 0) {
            sender.isSelected = false
            JAppViewTools.funj_showTextToast(self.view , message: String(format: "You can only choose %zd photos", JPhotosConfig.shared?.m_maxCountPhotos ?? 0))
            return
        }
        let index = Int((self.m_collectionView.contentOffset.x + 10 ) / self.m_collectionView.width)
        
        if index < self.m_dataArray.count {
            let model = self.m_dataArray[index] as? JPhotoPickerModel
            if sender.isSelected {
                model?.m_isSelected = true
                self.m_selectDataArr?.append(model!)
            } else {
                model?.m_isSelected = false
                self.m_selectDataArr?.removeAll(where: { (mo) -> Bool in
                    return (mo as? JPhotoPickerModel) == model
                })
            }
        }
        
    }
    func funj_getDetailInfo(_ saveImageDic : inout [AnyHashable : Any] , image : UIImage? , dic : [AnyHashable : Any]? , isDegraded : Bool , item : AVPlayerItem? , index : Int){
        if JPhotosConfig.shared?.m_currentIsVideo ?? false {
            if item != nil { saveImageDic["\(index)"] = item }
        } else {
            if image != nil { saveImageDic["\(index)"] = image }
        }
        if saveImageDic.count == self.m_selectDataArr!.count {
            self.funj_closeProgressView()
            var saveArr :[Any] = []
            for j in 0..<saveImageDic.count {
                if saveImageDic["\(j)"] != nil {
                    saveArr.append(saveImageDic["\(j)"] as Any)
                }
            }
            JPhotosConfig.shared?.m_selectCallback?(saveArr , JPhotosConfig.shared?.m_currentIsVideo ?? false )
            self.m_currentShowVCModel = .kCURRENTISPRENTVIEW
            self.modalTransitionStyle = .crossDissolve
            self.funj_clickBackButton()
        }
    }
    override func funj_clickBackButton(_ sender: UIButton? = nil) {
        self.m_changeCallback?(self.m_selectDataArr,self.m_isOrigalImage)
        super.funj_clickBackButton()
    }
}
extension JPhotosPreviewsVC {
    override func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return self.m_dataArray.count
    }
    override func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        return CGSize(width: collectionView.width, height: collectionView.height)
    }
    override func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: kCellIndentifier, for: indexPath) as! JPhotosPreviewsCell
        self.m_defaultImageView.isHidden = true
        if self.m_scrollIndex == -1 || indexPath.row == self.m_scrollIndex {
            self.m_scrollIndex = -1
            cell.funj_setBaseCollectionData(self.m_dataArray[indexPath.row])
        }
        return cell
    }
    override func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        for cell in self.m_collectionView.visibleCells {
            (cell as? JPhotosPreviewsCell)?.funj_stopAllPlayer()
        }
    }
    func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        let index = Int((scrollView.contentOffset.x + 10 ) / scrollView.width)
        let rightBt = self.navigationItem.rightBarButtonItem?.customView as? UIButton
        rightBt?.isSelected = false
        if index < self.m_dataArray.count {
            let model = self.m_dataArray[index] as? JPhotoPickerModel

            let isHas = self.m_selectDataArr?.contains(where: { (mo) -> Bool in
                return (mo as! JPhotoPickerModel) == model
            })
            if isHas ?? false {
                rightBt?.isSelected = model?.m_isSelected ?? false
            }
        }
    }
    func scrollViewDidEndScrollingAnimation(_ scrollView: UIScrollView) {
        scrollViewDidEndDecelerating(scrollView)
    }

}
